### ISSUE TITLE
MPI running errors in fbpic version of  0.16.0

### ISSUE THREAD

**Issue by QJohn2017** [fbpic/fbpic#453]
Hi, When I when I use MPI runing with CPU in fbpic, some errors occured:

- when I run this prgram **with 2 or more (e.g. 4) nodes**, there are always some errors:
`Traceback (most recent call last):
  File "modul_laser.py", line 131, in <module>
    n_order=n_order, use_cuda=use_cuda, verbose_level=2 )
  File "/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/lib/python3.7/site-packages/fbpic/main.py", line 324, in __init__
    print_simulation_setup( self, verbose_level=verbose_level )
  File "/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/lib/python3.7/site-packages/fbpic/utils/printing.py", line 209, in print_simulation_setup
    node_messages = sim.comm.mpi_comm.gather( node_message )
  File "mpi4py/MPI/Comm.pyx", line 1262, in mpi4py.MPI.Comm.gather
  File "mpi4py/MPI/msgpickle.pxi", line 680, in mpi4py.MPI.PyMPI_gather
  File "mpi4py/MPI/msgpickle.pxi", line 686, in mpi4py.MPI.PyMPI_gather
mpi4py.MPI.Exception: Unknown error class, error stack:
PMPI_Gatherv(416)...............: MPI_Gatherv failed(sbuf=0x2b858ef81940, scount=41, MPI_BYTE, rbuf=(nil), rcnts=(nil), displs=(nil), datatype=MPI_BYTE, root=0, comm=MPI_COMM_WORLD) failed
MPIR_Gatherv_impl(189)..........: Failure during collective
MPIR_Gatherv_impl(184)..........: 
MPIR_Gatherv_intra_auto(97).....: Failure during collective
MPIR_Gatherv_intra_auto(93).....: 
MPIR_Gatherv_allcomm_linear(159): 
MPIC_Send(144)..................: 
MPID_Send(73)...............
Traceback (most recent call last):
  File "modul_laser.py", line 146, in <module>
    add_laser_pulse( sim, synthesis_profile )
  File "/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/lib/python3.7/site-packages/fbpic/lpa_utils/laser/laser.py", line 79, in add_laser_pulse
    add_laser_direct( sim, laser_profile, boost )
  File "/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/lib/python3.7/site-packages/fbpic/lpa_utils/laser/direct_injection.py", line 66, in add_laser_direct
    local_array, with_damp=True)
  File "/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/lib/python3.7/site-packages/fbpic/boundaries/boundary_communicator.py", line 1071, in gather_grid_array
    self.mpi_comm.Gatherv( sendbuf, recvbuf, root=root )
  File "mpi4py/MPI/Comm.pyx", line 602, in mpi4py.MPI.Comm.Gatherv
mpi4py.MPI.Exception: Unknown error class, error stack:
PMPI_Gatherv(416)...............: MPI_Gatherv failed(sbuf=0x2b8c026e90d0, scount=197500, dtype=0x4c001041, rbuf=(nil), rcnts=(nil), displs=(nil), datatype=MPI_BYTE, root=0, comm=MPI_COMM_WORLD) failed
MPIR_Gatherv_impl(189)..........: Failure during collective
MPIR_Gatherv_impl(184)..........: 
MPIR_Gatherv_intra_auto(97).....: Failure during collective
MPIR_Gatherv_intra_auto(93).....: 
MPIR_Gatherv_allcomm_linear(159): 
MPIC_Send(144)..................: 
MPID_Send(73)...
yhrun: error: cn9956: tasks 2-3: Exited with exit code 1
yhrun: Job step aborted: Waiting up to 2 seconds for job step to finish.
yhrun: got SIGCONT
slurmd[cn9955]: *** STEP 14943215.0 CANCELLED AT 2020-06-07T09:26:41 ***
slurmd[cn9955]: *** JOB 14943215 CANCELLED AT 2020-06-07T09:26:41 ***
yhrun: forcing job termination
yhrun: error: cn9955: tasks 0-1: Terminated
`

- But if I **just use one node**, the running is ok:
`FBPIC (0.16.0)

MPI available: Yes
MPI processes used: 2
MPI Library Information: 
MPICH Version:	3.3.2
MPICH Release date:	Tue Nov 12 21:23:16 CST 2019
MPICH ABI:	13:8:1
MPICH Device:	ch3:nemesis
MPICH configure:	--prefix=/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3 --disable-dependency-tracking --enable-cxx --enable-fortran --disable-wrapper-rpath
MPICH CC:	x86_64-conda_cos6-linux-gnu-cc -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fdebug-prefix-map=/tmp/build/80754af9/mpich-mpi_1575396103289/work=/usr/local/src/conda/mpich-3.3.2 -fdebug-prefix-map=/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3=/usr/local/src/conda-prefix  -O2
MPICH CXX:	x86_64-conda_cos6-linux-gnu-c++ -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fvisibility-inlines-hidden -std=c++17 -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fdebug-prefix-map=/tmp/build/80754af9/mpich-mpi_1575396103289/work=/usr/local/src/conda/mpich-3.3.2 -fdebug-prefix-map=/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3=/usr/local/src/conda-prefix -O2
MPICH F77:	x86_64-conda_cos6-linux-gnu-gfortran -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fopenmp -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fdebug-prefix-map=/tmp/build/80754af9/mpich-mpi_1575396103289/work=/usr/local/src/conda/mpich-3.3.2 -fdebug-prefix-map=/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3=/usr/local/src/conda-prefix -O2
MPICH FC:	x86_64-conda_cos6-linux-gnu-gfortran -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fopenmp -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -I/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3/include -fdebug-prefix-map=/tmp/build/80754af9/mpich-mpi_1575396103289/work=/usr/local/src/conda/mpich-3.3.2 -fdebug-prefix-map=/WORK/ac_siom_jsliu2_1/qzy/App/miniconda3=/usr/local/src/conda-prefix -O2

CUDA available: No
Compute architecture: CPU
CPU multi-threading enabled: Yes
Threads: 12
FFT library: MKL
MPI rank 0 runs on node cn9669
MPI rank 1 runs on node cn9669

PSATD stencil order: 32
Particle shape: linear
Longitudinal boundaries: open
Transverse boundaries: reflective
Guard region size: 63 cells
Damping region size: 64 cells
Injection region size: 31 cells
Particle exchange period: every 14 step
Boosted frame: False

Initializing laser pulse on the mesh...
Done.

Just-In-Time compilation (up to one minute) ...[K
|                                   | 2/87599, calc. ETA..., 30539 ms/step[K
|                                   | 3/87599, calc. ETA..., 2568 ms/step[K
|                                   | 4/87599, calc. ETA..., 2576 ms/step[K
|                                   | 5/87599, calc. ETA..., 2579 ms/step[K
|                                   | 6/87599, calc. ETA..., 2577 ms/step[K
|                                   | 7/87599, calc. ETA..., 2606 ms/step[K
|                                   | 8/87599, calc. ETA..., 2567 ms/step[K
|                                   | 9/87599, calc. ETA..., 2590 ms/step[K
|                                   | 10/87599, calc. ETA..., 2560 ms/step[K
|                                   | 11/87599, calc. ETA..., 2578 ms/step[K
|                                   | 12/87599, calc. ETA..., 2570 ms/step[K
|                                   | 13/87599, calc. ETA..., 2575 ms/step[K
`

My submission script is :
`#!/bin/bash
export NUMBA_NUM_THREADS=12
export MKL_NUM_THREADS=12
yhrun -N 2 -n 4 --cpu_bind=sockets python test0.py
`
It seens that the errors is about mpi4py, But I reinstalled the mpi4py with pip for many times, the errors always occured. So, can any one help me? Thank you !

**Comment by RemiLehe** [comment]
Hi @QJohn2017,
Thanks for reporting this issue.

We have already observed in the past that the default version of `mpi4py` may not work on all clusters. Instead, on some clusters, you have to use a specialized version of `mpi4py`, which has been tuned specifically for that cluster's particular architecture.

See for instance the installation instructions for the cluster Summit:
https://fbpic.github.io/install/install_summit.html
In this case, instead of installing `mpi4py` by hand, we do `module load py-mpi4py`.

Another example is the installation instructions for the cluster Comet:
https://fbpic.github.io/install/install_comet.html
In this case, because the cluster does not provide a module for `mpi4py`, we had to install `mpi4py` by hand, but using a specific `mpicc` compiler.

So you can see from these examples that there is no easy, general solution for having `mpi4py` working on a given cluster, and that instead, each cluster will required a tailored solution.
Without having access to your cluster, it is difficult to tell you how to make it work. The best solution would be for you to ask your cluster administrator how to install/use `mpi4py` for this cluster.

**Comment by QJohn2017** [comment]
Hi, @RemiLehe , The MPI issues were resolved. Actually, When I use conda command to install mpi4py, the conda automally installed the mpich-3.3.2. But actually, the mpich-3.3.2 version is looks incompatible with mpi4py, which will lead the problrm of cross node operation. So, when I demote the mpich version to 3.2 then the issue was resolved. I have referenced the issue #391 that use the command `conda install -c conda-forge mpi4py=*=*mpich*`.
So, thank you very much, and this issue may be closed now.
